### PR TITLE
Seamless theme types in Kotlin/Native

### PR DESCRIPTION
**Pull request by justasm**
To share theme-aware UI component values across iOS and Android, we can move them to Kotlin/Native. The initial implementation in #25 moved primitive constants to K/N, but required mapping between primitives and Compose/SwiftUI-native types. Since the mapping had only to be done once, this was not an issue. It becomes an issue when we declare constants for specific UI components (see e.g. #34) as the mapping then has to be done inside of each component, on both iOS & Android.

The goal is to make all theme types - `TextStyle`, `Color` & `Dp` - interop as seamlessly as possible with SwiftUI and Jetpack Compose types even when shared via Kotlin Multiplatform, avoiding mapping where possible. Seamless `TextStyle` was added in #37. This PR adds seamless usage of `Color` and `Dp`.

There are two challenges with `Color` and `Dp` -
1. Compose `TextStyle` is a simple (`data`) `class`, which allows us to use `expect class TextStyle` and map to any Obj-C class (e.g. `UIFont`). Compose `Color` and `Dp`, on the other hand, are `inline class`es, which forces us to use `expect inline class` for seamless interop between K/N & Compose. We cannot `actual typealias` `inline class`es to any Obj-C class.
2. K/N has interop with Obj-C only, not with Swift. SwiftUI types - `Font`, `Color` and `CGFloat` - are Swift types, which means we cannot avoid additional mapping on the Swift side regardless.

As a result, we choose to achieve seamless interop between Kotlin/Android & Kotlin/Native, where it is possible, and present a solution for painless mapping in Swift.

#### Android
Using expect / actual, all three types map directly to Jetpack Compose types by the same name using `actual typealias`. This means no type mapping is required between the Android and Kotlin/Native bridge.

#### iOS
All three types are seamlessly mapped through a `@dynamicMemberLookup struct Kotlin<T>`. For example, let's look at #34 Kotlin constants for button and their types when accessing from Swift:
```kotlin
class ButtonConstants(theme: CurrentTheme) {
     val defaultColor = theme.colorPalette.primary                       // UInt64
     val defaultContentColor = theme.colorPalette.onPrimary              // UInt64
     val textStyle = theme.typographyScale.textL.copy(fontWeight = Bold) // UIFont
     val cornerRadius = theme.cornerRadiusScale.buttonRadius             // Float
 }
```
Using a `Kotlin` wrapper perfectly handles the mapping to types preferred by SwiftUI using key paths:
```swift
var constants: Kotlin<ButtonConstants> { Kotlin(ButtonConstants(theme: currentTheme)) }

constants.defaultColor        // Color
constants.defaultContentColor // Color
constants.textStyle           // Font
constants.cornerRadius        // CGFloat
```

### CODE REVIEW

File: android/ui/src/main/java/com/trafi/ui/theme/Color.kt [hunk 1]
```diff
@@ -45,22 +45,22 @@ fun MaasTheme.lightColors(
     onBackground = onBackground,
     onSurface = onSurface,
     onError = onError,
-    isLight = true
+    isLight = true,
```
Comment (eligijusTrafi) [inline android/ui/src/main/java/com/trafi/ui/theme/Color.kt:48]:
```suggestion
    isLight = true
```

File: android/ui/src/main/java/com/trafi/ui/theme/Color.kt [hunk 2]
```diff
@@ -45,22 +45,22 @@ fun MaasTheme.lightColors(
...
+    secondary: Color = ColorPalette.DefaultDark.Secondary,
+    secondaryVariant: Color = ColorPalette.DefaultDark.SecondaryVariant,
+    background: Color = ColorPalette.DefaultDark.Background,
+    surface: Color = ColorPalette.DefaultDark.Surface,
+    error: Color = ColorPalette.DefaultDark.Error,
+    onPrimary: Color = ColorPalette.DefaultDark.OnPrimary,
+    onSecondary: Color = ColorPalette.DefaultDark.OnSecondary,
+    onBackground: Color = ColorPalette.DefaultDark.OnBackground,
+    onSurface: Color = ColorPalette.DefaultDark.OnSurface,
+    onError: Color = ColorPalette.DefaultDark.OnError,
```
Comment (eligijusTrafi) [inline android/ui/src/main/java/com/trafi/ui/theme/Color.kt:63]:
```suggestion
    onError: Color = ColorPalette.DefaultDark.OnError
```

File: android/ui/src/main/java/com/trafi/ui/theme/Color.kt [hunk 3]
```diff
@@ -74,7 +74,7 @@ fun MaasTheme.darkColors(
     onBackground = onBackground,
     onSurface = onSurface,
     onError = onError,
-    isLight = false
+    isLight = false,
```
Comment (eligijusTrafi) [inline android/ui/src/main/java/com/trafi/ui/theme/Color.kt:77]:
```suggestion
    isLight = false
```

File: android/ui/src/main/java/com/trafi/ui/theme/Color.kt [hunk 4]
```diff
@@ -91,7 +91,7 @@ class MaasColorPalette(
     onBackground: Color,
     onSurface: Color,
     onError: Color,
-    isLight: Boolean
+    isLight: Boolean,
```
Comment (eligijusTrafi) [inline android/ui/src/main/java/com/trafi/ui/theme/Color.kt:94]:
```suggestion
    isLight: Boolean
```

File: android/ui/src/main/java/com/trafi/ui/theme/CurrentTheme.kt [hunk 5]
```diff
@@ -12,52 +9,43 @@ import com.trafi.ui.theme.internal.CurrentTypographyScale
...
+                secondary = secondary,
+                secondaryVariant = secondaryVariant,
+                background = background,
+                surface = surface,
+                error = error,
+                onPrimary = onPrimary,
+                onSecondary = onSecondary,
+                onBackground = onBackground,
+                onSurface = onSurface,
+                onError = onError,
```
Comment (eligijusTrafi) [inline android/ui/src/main/java/com/trafi/ui/theme/CurrentTheme.kt:26]:
```suggestion
                onError = onError
```

File: android/ui/src/main/java/com/trafi/ui/theme/CurrentTheme.kt [hunk 6]
```diff
@@ -12,52 +9,43 @@ import com.trafi.ui.theme.internal.CurrentTypographyScale
...
+        typographyScale = with(MaasTheme.typography) {
+            CurrentTypographyScale(
+                headingXXL = headingXXL,
+                headingXL = headingXL,
+                headingL = headingL,
+                headingM = headingM,
+                textL = textL,
+                textM = textM,
+                textS = textS,
+                label = label,
```
Comment (eligijusTrafi) [inline android/ui/src/main/java/com/trafi/ui/theme/CurrentTheme.kt:38]:
```suggestion
                label = label
```

File: common/core/src/androidMain/kotlin/com/trafi/ui/theme/internal/TextStyle.kt [hunk 7]
```diff
@@ -1,30 +1,29 @@
...
 
-actual typealias OsTextStyle = TextStyle
+actual typealias TextStyle = androidx.compose.ui.text.TextStyle
 
-internal actual fun OsTextStyle.copy(
+internal actual fun TextStyle.copy(
     fontStyle: BasicFontStyle?,
     fontWeight: BasicFontWeight?,
     fontSize: Int?,
     lineHeight: Int?,
```
Comment (eligijusTrafi) [inline common/core/src/androidMain/kotlin/com/trafi/ui/theme/internal/TextStyle.kt:13]:
```suggestion
    lineHeight: Int?
```

File: common/core/src/androidMain/kotlin/com/trafi/ui/theme/internal/TextStyle.kt [hunk 8]
```diff
@@ -1,30 +1,29 @@
...
     fontSize = fontSize?.sp ?: this.fontSize,
     lineHeight = lineHeight?.sp ?: this.lineHeight
 )
 
-internal actual fun OsTextStyle(
+internal actual fun TextStyle(
     fontStyle: BasicFontStyle,
     fontWeight: BasicFontWeight,
     fontSize: Int,
     lineHeight: Int,
```
Comment (eligijusTrafi) [inline common/core/src/androidMain/kotlin/com/trafi/ui/theme/internal/TextStyle.kt:25]:
```suggestion
    lineHeight: Int
```

File: common/core/src/commonMain/kotlin/com/trafi/ui/theme/internal/CurrentColorPalette.kt [hunk 9]
```diff
@@ -1,16 +1,16 @@
...
+    val secondary: Color,
+    val secondaryVariant: Color,
+    val background: Color,
+    val surface: Color,
+    val error: Color,
+    val onPrimary: Color,
+    val onSecondary: Color,
+    val onBackground: Color,
+    val onSurface: Color,
+    val onError: Color,
```
Comment (eligijusTrafi) [inline common/core/src/commonMain/kotlin/com/trafi/ui/theme/internal/CurrentColorPalette.kt:15]:
```suggestion
    val onError: Color
```

File: common/core/src/commonMain/kotlin/com/trafi/ui/theme/internal/CurrentCornerRadiusScale.kt [hunk 10]
```diff
@@ -1,5 +1,5 @@
 package com.trafi.ui.theme.internal
 
 class CurrentCornerRadiusScale(
-    val buttonRadius: OsDimension,
+    val buttonRadius: Dp,
```
Comment (eligijusTrafi) [inline common/core/src/commonMain/kotlin/com/trafi/ui/theme/internal/CurrentCornerRadiusScale.kt:4]:
```suggestion
    val buttonRadius: Dp
```

File: common/core/src/commonMain/kotlin/com/trafi/ui/theme/internal/CurrentSpacingScale.kt [hunk 11]
```diff
@@ -1,5 +1,5 @@
 package com.trafi.ui.theme.internal
 
 class CurrentSpacingScale(
-    val globalMargin: OsDimension,
+    val globalMargin: Dp,
```
Comment (eligijusTrafi) [inline common/core/src/commonMain/kotlin/com/trafi/ui/theme/internal/CurrentSpacingScale.kt:4]:
```suggestion
    val globalMargin: Dp
```

File: common/core/src/commonMain/kotlin/com/trafi/ui/theme/internal/CurrentTypographyScale.kt [hunk 12]
```diff
@@ -1,12 +1,12 @@
...
-    val textS: OsTextStyle,
-    val label: OsTextStyle,
+    val headingXXL: TextStyle,
+    val headingXL: TextStyle,
+    val headingL: TextStyle,
+    val headingM: TextStyle,
+    val textL: TextStyle,
+    val textM: TextStyle,
+    val textS: TextStyle,
+    val label: TextStyle,
```
Comment (eligijusTrafi) [inline common/core/src/commonMain/kotlin/com/trafi/ui/theme/internal/CurrentTypographyScale.kt:11]:
```suggestion
    val label: TextStyle
```

File: common/core/src/commonMain/kotlin/com/trafi/ui/theme/internal/Dsl.kt [hunk 13]
```diff
@@ -11,7 +11,7 @@ internal class TextStyleBuilder(
     var fontSize: Int? = null,
     var lineHeight: Int? = null,
```
Comment (eligijusTrafi) [inline common/core/src/commonMain/kotlin/com/trafi/ui/theme/internal/Dsl.kt:12]:
```suggestion
    var lineHeight: Int? = null
```

File: common/core/src/commonMain/kotlin/com/trafi/ui/theme/internal/TextStyle.kt [hunk 14]
```diff
@@ -1,17 +1,17 @@
...
 
-expect class OsTextStyle
+expect class TextStyle
 
-internal expect fun OsTextStyle.copy(
+internal expect fun TextStyle.copy(
     fontStyle: BasicFontStyle? = null,
     fontWeight: BasicFontWeight? = null,
     fontSize: Int? = null,
     lineHeight: Int? = null,
```
Comment (eligijusTrafi) [inline common/core/src/commonMain/kotlin/com/trafi/ui/theme/internal/TextStyle.kt:9]:
```suggestion
    lineHeight: Int? = null
```

File: common/core/src/commonMain/kotlin/com/trafi/ui/theme/internal/TextStyle.kt [hunk 15]
```diff
@@ -1,17 +1,17 @@
...
-): OsTextStyle
+): TextStyle
 
-internal expect fun OsTextStyle(
+internal expect fun TextStyle(
     fontStyle: BasicFontStyle,
     fontWeight: BasicFontWeight,
     fontSize: Int,
     lineHeight: Int,
```
Comment (eligijusTrafi) [inline common/core/src/commonMain/kotlin/com/trafi/ui/theme/internal/TextStyle.kt:16]:
```suggestion
    lineHeight: Int
```

**Review by eligijusTrafi — APPROVED** [review]
👏

File: android/ui/src/main/java/com/trafi/ui/theme/Color.kt (hunk 1 above)
Comment (justasm) [reply android/ui/src/main/java/com/trafi/ui/theme/Color.kt:48]: This was intentional 🙂 Kotlin 1.4 supports trailing parameter commas. The primary benefits are
- if we add a new parameter at the end, this line won't change in git
- easier to rearrange lines if you want to change the parameter order

**Pull request merged**